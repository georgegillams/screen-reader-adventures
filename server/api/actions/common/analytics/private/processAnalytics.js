const groupingProperties = ['browser', 'os', 'url', 'utm_source', 'utm_medium', 'ipAddressPrefix', 'browserVersion'];

/**
 * Creates a key for a data point
 * @param {object} dataPoint The data point to consider
 * @returns {string} The key for the data point
 */

const keyForData = dataPoint => {
  const values = [];
  groupingProperties.forEach(gp => {
    values.push(dataPoint[gp]);
  });
  return values.join('_');
};

/**
 * Inserts the new data point into the list, or if it exists increments the count for the matching entry
 * @param {array} list The list to update
 * @param {object} dataPoint The data point to consider
 * @returns {null} none
 */

const insertOrUpdateCount = (list, dataPoint) => {
  const dataPointKey = keyForData(dataPoint);
  const matchingEntries = list.filter(l => l.key === dataPointKey);
  if (matchingEntries.length > 0) {
    matchingEntries[0].count += 1;
  } else {
    const newValue = { key: dataPointKey, count: 1 };
    groupingProperties.forEach(gp => {
      newValue[gp] = dataPoint[gp];
    });
    list.push(newValue);
  }
};

/**
 * Processes analytic data to group analytics by characteristics and count instances of each
 * @param {array} data The data point to consider
 * @returns {array} The unique analytics with counts of how many times they occur
 */

const processAnalytics = data => {
  const uniqueDataPoints = [];
  data.forEach(d => {
    insertOrUpdateCount(uniqueDataPoints, d);
  });
  return uniqueDataPoints;
};

export default processAnalytics;
