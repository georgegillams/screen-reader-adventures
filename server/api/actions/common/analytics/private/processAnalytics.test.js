#!/usr/bin/env node

import processAnalytics from './processAnalytics.js';

test('process analytics data properly ', done => {
  const data = [
    { ipAddressPrefix: '4.1.X.X', browser: 'browser1', browserVersion: '1', os: 'os1', url: '/url/1' },
    { ipAddressPrefix: '4.1.X.X', browser: 'browser1', browserVersion: '1', os: 'os2', url: '/url/2' },
    { ipAddressPrefix: '4.1.X.X', browser: 'browser1', browserVersion: '1', os: 'os1', url: '/url/1' },
    {
      ipAddressPrefix: '1.2.X.X',
      browser: 'browser2',
      browserVersion: '2',
      os: 'os2',
      url: '/url/4',
      utm_source: 'utmSource1',
      utm_medium: 'utmMedium1',
    },
    { ipAddressPrefix: '1.2.X.X', browser: 'browser2', browserVersion: '2', os: 'os2', url: '/url/4' },
    { ipAddressPrefix: '1.2.X.X', browser: 'browser2', browserVersion: '2', os: 'os1', url: '/url/1' },
    { ipAddressPrefix: '1.0.X.X', browser: 'browser2', browserVersion: '2', os: 'os2', url: '/url/1' },
    {
      ipAddressPrefix: '1.0.X.X',
      browser: 'browser2',
      browserVersion: '2',
      os: 'os2',
      url: '/url/1',
      someUnknownProperty: 'ðŸ‘½',
    },
    { ipAddressPrefix: '1.2.X.X', browser: 'browser3', browserVersion: '3', os: 'os1', url: '/url/6' },
    { ipAddressPrefix: '1.2.X.X', browser: 'browser4', browserVersion: '4', os: 'os3', url: '/url/1' },
    { ipAddressPrefix: '1.2.X.X', browser: 'browser1', browserVersion: '1', os: 'os1', url: '/url/1' },
    { ipAddressPrefix: '1.2.X.X', browser: 'browser1', browserVersion: '1', os: 'os1', url: '/url/1' },
    { ipAddressPrefix: '1.2.X.X', browser: 'browser1', browserVersion: '1', os: 'os1', url: '/url/1' },
    { ipAddressPrefix: '1.2.X.X', browser: 'browser1', browserVersion: '1', os: 'os1', url: '/url/1' },
    { ipAddressPrefix: '1.2.X.X', browser: 'browser1', browserVersion: '1.3', os: 'os2', url: '/url/2' },
    { ipAddressPrefix: '1.2.X.X', browser: 'browser1', browserVersion: '1.3', os: 'os2', url: '/url/2' },
    { ipAddressPrefix: '1.2.X.X', browser: 'browser1', browserVersion: '1.3', os: 'os2', url: '/url/2' },
    { ipAddressPrefix: '1.2.X.X', browser: 'browser1', browserVersion: '1.3', os: 'os2', url: '/url/2' },
  ];

  const result = processAnalytics(data);
  expect(result).toMatchSnapshot();

  done();
});
