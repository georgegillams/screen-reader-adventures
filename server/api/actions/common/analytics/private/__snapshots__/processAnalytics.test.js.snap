// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`process analytics data properly  1`] = `
Array [
  Object {
    "browser": "browser1",
    "browserVersion": "1",
    "count": 2,
    "ipAddressPrefix": "4.1.X.X",
    "key": "browser1_os1_/url/1___4.1.X.X_1",
    "os": "os1",
    "url": "/url/1",
    "utm_medium": undefined,
    "utm_source": undefined,
  },
  Object {
    "browser": "browser1",
    "browserVersion": "1",
    "count": 1,
    "ipAddressPrefix": "4.1.X.X",
    "key": "browser1_os2_/url/2___4.1.X.X_1",
    "os": "os2",
    "url": "/url/2",
    "utm_medium": undefined,
    "utm_source": undefined,
  },
  Object {
    "browser": "browser2",
    "browserVersion": "2",
    "count": 1,
    "ipAddressPrefix": "1.2.X.X",
    "key": "browser2_os2_/url/4_utmSource1_utmMedium1_1.2.X.X_2",
    "os": "os2",
    "url": "/url/4",
    "utm_medium": "utmMedium1",
    "utm_source": "utmSource1",
  },
  Object {
    "browser": "browser2",
    "browserVersion": "2",
    "count": 1,
    "ipAddressPrefix": "1.2.X.X",
    "key": "browser2_os2_/url/4___1.2.X.X_2",
    "os": "os2",
    "url": "/url/4",
    "utm_medium": undefined,
    "utm_source": undefined,
  },
  Object {
    "browser": "browser2",
    "browserVersion": "2",
    "count": 1,
    "ipAddressPrefix": "1.2.X.X",
    "key": "browser2_os1_/url/1___1.2.X.X_2",
    "os": "os1",
    "url": "/url/1",
    "utm_medium": undefined,
    "utm_source": undefined,
  },
  Object {
    "browser": "browser2",
    "browserVersion": "2",
    "count": 2,
    "ipAddressPrefix": "1.0.X.X",
    "key": "browser2_os2_/url/1___1.0.X.X_2",
    "os": "os2",
    "url": "/url/1",
    "utm_medium": undefined,
    "utm_source": undefined,
  },
  Object {
    "browser": "browser3",
    "browserVersion": "3",
    "count": 1,
    "ipAddressPrefix": "1.2.X.X",
    "key": "browser3_os1_/url/6___1.2.X.X_3",
    "os": "os1",
    "url": "/url/6",
    "utm_medium": undefined,
    "utm_source": undefined,
  },
  Object {
    "browser": "browser4",
    "browserVersion": "4",
    "count": 1,
    "ipAddressPrefix": "1.2.X.X",
    "key": "browser4_os3_/url/1___1.2.X.X_4",
    "os": "os3",
    "url": "/url/1",
    "utm_medium": undefined,
    "utm_source": undefined,
  },
  Object {
    "browser": "browser1",
    "browserVersion": "1",
    "count": 4,
    "ipAddressPrefix": "1.2.X.X",
    "key": "browser1_os1_/url/1___1.2.X.X_1",
    "os": "os1",
    "url": "/url/1",
    "utm_medium": undefined,
    "utm_source": undefined,
  },
  Object {
    "browser": "browser1",
    "browserVersion": "1.3",
    "count": 4,
    "ipAddressPrefix": "1.2.X.X",
    "key": "browser1_os2_/url/2___1.2.X.X_1.3",
    "os": "os2",
    "url": "/url/2",
    "utm_medium": undefined,
    "utm_source": undefined,
  },
]
`;
